enterprise/backend/test/metabase_enterprise/caching/strategies_test.clj

```diff
@@ -0,0 +1,257 @@
+
+      (let [query (mt/mbql-query venues {:order-by [[:asc $id]] :limit 5})
+            mkres (fn [input]
+                    {:cache/details (if input
+                                      {:cached true, :updated_at input, :hash some?}
+                                      {:stored true, :hash some?})
+                     :row_count     5
+                     :status        :completed})]
+        (mt/with-model-cleanup [[:model/QueryCache :updated_at]]
+          (testing "strategy = ttl"
+            (let [query (assoc query :cache-strategy {:type           :ttl
+                                                      :multiplier     10
+                                                      :min_duration   0
+                                                      :avg-execution-ms 500})]
+              (testing "Results are stored and available immediately"
+                (mt/with-clock #t "2024-02-13T10:00:00Z"
+                  (is (=? (mkres nil)
+                          (-> (qp/process-query query) (dissoc :data))))
+                  (is (=? (mkres #t "2024-02-13T10:00:00Z")
+                          (-> (qp/process-query query) (dissoc :data))))))
+              (testing "4 seconds past that results are still there - 10 * 500 = 5 seconds"
+                (mt/with-clock #t "2024-02-13T10:00:04Z"
+                  (is (=? (mkres #t "2024-02-13T10:00:00Z")
+                          (-> (qp/process-query query) (dissoc :data))))))
+              (testing "6 seconds later results are unavailable"
+                (mt/with-clock #t "2024-02-13T10:00:06Z"
+                  (is (=? (mkres nil)
+                          (-> (qp/process-query query) (dissoc :data)))))))))
+
+        (mt/with-model-cleanup [[:model/QueryCache :updated_at]]
+          (testing "strategy = duration"
+            (let [query (assoc query :cache-strategy {:type         :duration
+                                                      :duration     1
+                                                      :unit         "minutes"
+                                                      :min_duration 0})]
+              (testing "Results are stored and available immediately"
+                (mt/with-clock #t "2024-02-13T10:00:00Z"
+                  (is (=? (mkres nil)
+                          (-> (qp/process-query query) (dissoc :data))))
+                  (is (=? (mkres #t "2024-02-13T10:00:00Z")
+                          (-> (qp/process-query query) (dissoc :data)))))
+                (mt/with-clock #t "2024-02-13T10:00:59Z"
+                  (is (=? (mkres #t "2024-02-13T10:00:00Z")
+                          (-> (qp/process-query query) (dissoc :data)))))
+                (mt/with-clock #t "2024-02-13T10:01:01Z"
+                  (is (=? (mkres nil)
+                          (-> (qp/process-query query) (dissoc :data)))))))))
+
+        (mt/with-model-cleanup [[:model/QueryCache :updated_at]]
+          (testing "strategy = schedule || query"
```

Comment:
Even though the config of a "query" config is a superset of a "schedule", and the behavior hinges on the same expiry timestamp, I think its worth having an explicit separate test case.

Reply:
Agree, done in b3d0e32fcf